### ISSUE TITLE
Auto Update of UN OCHA data

### ISSUE THREAD

**Issue by MichaelBrandonFalk** [wmgeolab/geoBoundaries#2295]
Hello,

There are several countries missing info or with outdated data in the UN OCHA column, but I am able to find the data on humdata. Would it be possible for Geoboundaries to periodically pull from humdata ( maybe monthly). This would resolve the Pakistan issue whish is now tagged as Urgent. It would also update many other countries including Rwanda (OCHA's layer is really good for Rwanda) , Ethiopia, and many other countries. Thank you.

**Comment by DanRunfola** [comment]
Hi MIchael,

We've been debating the value of mirroring HDX for a little while now, though this is tentatively on the list if we do.  If possible, could you provide a little more detail on your use case?  I.e., why do you use geoBoundaries to access the UN OCHA data, as contrasted to humdata?

-Dan

**Comment by leeberryman** [comment]
@DanRunfola, @MichaelBrandonFalk is one of the stakeholders in consuming a dataset that we have built downstream as a pipeline from geoboundaries, geonames, OSM hamlets and age/sex pop grid values from Worldpop 100m.  At this point we only pull from gBOpen, but all of our stakeholders are non profits and the gBHumanitarian release is suitable for their needs, so @MichaelBrandonFalk has pointed out that in many cases the OCHA data is more recent than gBOpen and in several cases OCHA has newer data than gBHumanitarian.  Our pipeline is already built out to get all the appropriate source and attribution data from the GeoBoundaries API, if their are more development hours needed to make the mirror happen I'm happy to talk through what needs to happen so that we could meet everyone's needs.

**Comment by DanRunfola** [comment]
Got it - and, yes, we would anticipate the OCHA data to be ahead of ours in many cases (with the tradeoff being licensure).

@karimbahgat built a web interface and hooks that automatically generate pull requests for us to process, which I think would be the appropriate way to do this (i.e., a nightly cron --> new HDX file identified --> Pull request generated --> we pull it in after a check).  He can provide more details on the API, but you can also see it in action here: https://www.geoboundaries.org/gbContribute.html

behind the scenes for the hookds: https://github.com/wmgeolab/gbWeb/blob/main/gbContribute.html

**Comment by DanRunfola** [comment]
(and the nice thing about that API is that it's something anyone can ping against - i.e., we don't need to host the chron; any user could).

**Comment by leeberryman** [comment]
Thanks @DanRunfola I'll create a work item for one of our team members to look at this and they might reach out as they have questions.

**Comment by maxmalynowsky** [comment]
@leeberryman @MichaelBrandonFalk, I also maintain a project that aggregates UN OCHA and geoBoundaries data together. This might be useful for the use case you describe, it's updated weekly with new UN OCHA data from HDX.

- https://fieldmaps.io/data

Pulling data directly from HDX still involves manual cleaning of file and field names, sometimes also topology cleaning too, so I've made the intermediary files of my project public so others don't have to repeat this work. It even has metadata available as JSON or CSV that a script can use to detect changes against a last modified date: 

- https://fieldmaps.io/data/cod
- https://data.fieldmaps.io/cod.json
- https://data.fieldmaps.io/cod.csv

**Comment by MichaelBrandonFalk** [comment]
@maxmalynowsky Very cool! Thanks for sharing!

**Comment by DanRunfola** [comment]
Absolutely!  As you can probably tell, until our new hire comes onboard I'm pretty swamped to help with new intiiatives, but in about a month or so (say, post 5.0) we can probably help.  Might be a good first or second project for someone on our side as well.

**Comment by maxmalynowsky** [comment]
@DanRunfola @leeberryman , I had written some code a while back that converted OCHA to geoBoundaries format. I just uploaded the outputs of that to the following location which can be downloaded and placed directly in [gbHumanitarian](https://github.com/wmgeolab/geoBoundaries/tree/main/sourceData/gbHumanitarian) without any modification. I created some metadata files to make it easier to manage this, and I've set up my pipeline to update these outputs at the same time I generate outputs for FieldMaps.

- https://data.fieldmaps.io/gb-humanitarian.json
- https://data.fieldmaps.io/gb-humanitarian.csv
- https://data.fieldmaps.io/gb-humanitarian.xlsx

**Comment by Raygone-pwd** [comment]
@DanRunfola can we extend /gbContribute to have a field or option to specify whether the upload data should go to gbOpen, gbHumanitarian or gbAuthoritative.

**Comment by maxmalynowsky** [comment]
I saw there was some discussion on that here: https://github.com/wmgeolab/geoBoundaries/discussions/2342

I agree with Dan's view that those streams should be mirrors that automatically hook into pipelines from those sources, and don't allow individual user submissions.

**Comment by DanRunfola** [comment]
@Raygone2 @maxmalynowsky Yes, agreed with my previous thought - essentially, we have three streams now in gB:
1) gbOpen - This is the community-driven, open license gB release.  We receive all sorts of data from all sorts of people on this.
2) gbHumanitarian - This is a mirror of UN OCHA's CODs, standardized to the gB schema.  We don't expect misc. community contributions here, as it's just a mirror we need to program, so optimally (and hopefully in the near future) won't come in through the gbContribute tool at all.
3) gbAuthoritative - This is a mirror of UN SALB's layers, standardized to the gB schema.  Just like the OCHA case, we have a single source, so a programmatic pull is the goal (I have a studnet working on this one now).

Right now the use of gbContribute for Humanitarian cases is a bit of a crutch, until we get the automated systems online.  I'm in the middle of a hire now, who will hopefully be able to focus on this pipeline pretty quick after we get them here.

**Comment by Raygone2** [comment]
> @maxmalynowsky thanks! @DanRunfola are you interested in us enhancing @karimbahgat's pull request process from @maxmalynowsky's cleaned up CODs (https://data.fieldmaps.io/cod.json) and making pull requests to gBHumanitarian?

I was trying to work on this. But I realized there is no way to differentiate the data (gbOpen or gbHumanitarian). Thus, my question; if there is anything that could be done for this process at the moment.

**Comment by leeberryman** [comment]
@DanRunfola I have asked @RayGone to look at pulling in the most recent UN OCHA data to gBHumanitarian as the changes happen.  I think using @maxmalynowsky's weekly cleaned up  versions would be great, but if you tell us that the raw updates is what is desired I understand.  @DanRunfola also if this isn't something you want help on the pipeline please tell us.

**Comment by DanRunfola** [comment]
Ok - now I'm starting to see the broader issue.

Basically, if we want an optimal workflow to be automated PR --> human approval (i.e., not fully programmatic, but nearly), then we need some tool to generate the PRs.  What Jim Courson has implemented now is a part of that, but he's (a) stuck with submitting to gbOpen (as that's what our API allows for right now), and (b) doesn't have the file format for the zip submission right, so everything is failing (i.e., see https://github.com/wmgeolab/geoBoundaries/pull/2349).

So, if we want to move the ball on that, what I would ask is that @Raygone2 first work on improving the PR submission script that Jim has built (not sure where that lives) so that the files actually pass the github bot checks.  Then, in parallel we'll update the API to allow for pushes into the gbHumanitarian dataset, and meet in the middle.

Does that work @Raygone2 ? Also @MichaelBrandonFalk

**Comment by fititnt** [comment]
1. https://en.wikipedia.org/wiki/Common_Operational_Datasets
2. https://cod.unocha.org/
3. https://data.humdata.org/dataset?ext_cod=1
4. https://en.wikipedia.org/wiki/Place_code
5. https://en.wikipedia.org/wiki/Authority_control

---

I agree with max and dans response. Anyway, let me give more context, so people trying to help geoBoundaries (but not sure what is used) move from the idea of "allow anyone submit data" to maybe help to automate things (or fix what break the pipelines) but still keep faithful for OCHA CODs (e.g. change data format, maybe schema, but consider that P-Codes are used for data exchange).

## About OCHA's COD-ABs

For who's not aware, by "humanitarian", its part (actually the most used one from what is called Common Operational Datasets (or CODs) [1] [2] which is also available on a open data platform powered by CKAN [3].

The code to match the boundaries with other data is called P-Code [4]. Which means that UN OCHA act as a authority control [5] for these P-Codes (which go in more details than first level), and beyond the boundaries themselves, it also provide natural language labels for the places.

### What is special about COD-ABs

While at first you here might see the licensing issues, actually COD-ABs fill a gap of "neutral spatial referential codes" good enough to be used on humanitarian settings, where people will not agree. This is actually pretty bad, because parties in a conflict or donors aiding a conflict would have strong opinions. The way P-Codes are designed, this means OCHA and others will also strongly focus on use these P-Codes as pivot code for other data. The most obvious example are pre-compiled list of population statistics by each P-code. However, since is not feasible for humanitarian IGOs (here at international level) develop all potential software anyone in the ground could use, in addition to make data interoperable, another challenge is that if someone try to use a software (let's say, using public domain, geoBoundaries, etc, something not designed to be sensitive to disputes, which can happens with data from OpenStreetMap) and plot the data into a map, the following might happen:

- **_"I don't like the name of the place on this map, use something else or I will ignore what you have to say"._ (in other words: no matter how great a solution is, the entire software or tooling might be rejected)**

This explain why, while anyone could take some public domain administrative boundary and call it "this is for humanitarian use" and I'm very sure this is with good intent, even if geoBoundaries does not state that they are based on OCHA COD-ABs and would use generic terms such as "humanitarian", there's a strong motivation to have an more predictable way to make such auto update without average end user intervention.

# Special case about need to fast updates

**Another important aspect on CODs is also make as easier as possible to avoid end users not update the data they already have**. However the way CODs are used, they tend to be distributed near the software used by the users. Something such as geoBoundaries is one level above (because work as distribution channel, with know schema), but myself I do have needs close to application such as on https://github.com/biolab/orange3-geo/issues/149.

The big picture is that theirs an entire chain of how the boundaries can be be conversed to something users need, and the less likely to make data happens, the better.

## However, the challenge of have automated updates is not new

**Being very, very frankly here, the COD-ABs data shared on [1] from time to time will break whatever automation do exist, even for who know what is going on**. Note that even if CODs got more documented schema, it can still have edge cases like for political reasons, but we still have issues with data formatting. The end result is the following: even if end users receive complains about they map or printed PDF be refused, who develop solutions (again, often not paid for this, doing pro bono publico) **for them might simply not want to keep fixing their previous scripts because something new happened**.

While I understand people here might already think that geoBoundaries might not be in sync with OCHA CODs (let's say, months), the reality is likely even humanitarian agencies either **using years old CODs** or using commercial boundaries because people do not want to keep doing data cleaning the original sources.

I really like the idea about have more people to do "human in the loop", but the big picture is both repackaging of COD-ABs (like geoBoundaries with predictable schema) or software which needs to ship cached version of COD-ABs will be used in production. For things already not automated (or if a script breaks), I still think that **redundancy** on the human side is important o take in account upfront (as is the case when re-packaging), but for applications, the general idea is educate developers to allow users re-fetch updated versions, but even these developers would need to have something such as geoBoundaries or fieldmaps).

**Comment by DanRunfola** [comment]
@fititnt Many good points here, and I agree strongly on the redundancy side.

I think the proposal on the table is to have @leeberryman work with his team (@Raygone2 ) to take the script that's currently creating PRs from HDX, improve that script so the PRs coming in are "ready to go", and then we have our humans approve the PRs.  If something breaks, we'll end up with nonsense data we would still manually correct, but with an obvious incentive to fix the script itself.

On redundancy, this is a hard one, but the way I approach it for gB is through student's that are interested in getting engaged on a research project.  That's why we're so active during the semesters and so quiet during the summers :).

**Comment by maxmalynowsky** [comment]
I just forked the repo, and until a script is built that creates pull requests from HDX, I wouldn't consider it to be much extra effort to create manual pull requests while I'm updating data for FieldMaps.io. Something to consider when attempting automation from HDX as a source is that only a minority of datasets there adhere to a strict attribute schema. These are about 50 or so treated as "Enhanced CODs" managed by the University of Georgia. The remaining 100+ datasets vary widely in individual formatting.

Also, for all data on HDX, files are uniquely named by source and date, meaning that any script would break as soon as there's a new update. For example, if you take a URL like https://data.humdata.org/dataset/cod-ab-lka, it's 2022 Geodatabase was named `lka_adm_20220802.gdb.zip`, which was later updated to `lka_adm_20220816.gdb.zip` after a minor correction was made. This update pattern isn't consistent throughout HDX, at https://data.humdata.org/dataset/cod-ab-gmb the Geodatabase went from being `gmb_adm_2022.gdb.zip` to `gmb_adm_2022_v2.gdb.zip` to fix a correction.

I don't see any way of writing a script around this uncertainty, so at the moment I manually download and standardize every file from HDX, giving it a human review by myself before it goes into my own project. Since I'm already doing all this work, feel free to take advantage of it since duplicating this would likely be quite labour intensive.

**Comment by DanRunfola** [comment]
@maxmalynowsky You're a wizard.

Give me a few days (probably next week) to get that ingested - hope is to clear out the current crop of PRs before end-of-Sept for a 5.0 launch.  Hopefully right after that I'll have a new hire start that can help with some of the mountain of refinements we have in issues as well.

**Comment by DanRunfola** [comment]
Well, in a very timely fashion I am just about to push this, a good three months later than I thought :)

But, good news is the entire build is nearly rewritten for a HPC workflow, which shoudl facilitate nightly builds (around an hour on 40 cores across 20 nodes right now - memory is the only real bottleneck, and I can scale up to whatever we need!).

Not closing this as I'm still not at the auto-pulls of OCHA, but one step forward :)
